### PR TITLE
Adds support for retrieving metadata from Skuid Platform

### PR DESCRIPTION
**Pull request by zachelrath**
- Adds login / connection utilities for Skuid Platform
- Adds new command `skuid retrieve` which currently only supports retrieving all metadata because that's all the API supports (for now, but Ben is working on that...)
- Adjusts documentation and command API's to support both Platforms

### CODE REVIEW

File: pliny/pliny.go [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+	if verboseLogs {
+		fmt.Println("Login successful! Access Token: " + conn.AccessToken)
+	}
+
+	return api, err
+}
+
+// Used to obtain an OAuth2 access_token
+func (conn *RestConnection) Refresh() (err error) {
```
Comment (micahhausler) [inline pliny/pliny.go]: Since you're initializing `err` below in your `resp, err := http.PostForm`, you don't strictly _need_ to define `(err error)` for your return. You could simply do `Refresh() error {`. No change required: but just an FYI.

File: pliny/pliny.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+	if params != nil && len(params) != 0 {
+		endpoint += "?" + params.Encode()
+	}
+
+	var body io.Reader
+
+	if payload != nil {
+		jsonBytes, err := json.Marshal(payload)
+
+		// quoted := []byte(strconv.Quote(string(jsonBytes)))
```
Comment (micahhausler) [inline pliny/pliny.go]: Can you remove this comment?

File: pliny/pliny.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+	return conn.MakeRequest("GET", url, nil, query)
+}
+
+// Makes a POST request
+func (conn *RestConnection) Post(url string, payload interface{}) (result []byte, err error) {
+	return conn.MakeRequest("POST", url, payload, nil)
+}
+
+// Executes an HTTP request
+func (conn *RestConnection) MakeRequest(method string, url string, payload interface{}, params url.Values) (result []byte, err error) {
```
Comment (micahhausler) [inline pliny/pliny.go]: Since payload has to be marshalable to JSON, can you make payload a [Marshaler](https://golang.org/pkg/encoding/json/#Marshaler)?

```go
func (conn *RestConnection) MakeRequest(method string, url string, payload json.Marshaler, params url.Values)
```

File: pliny/pliny.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+	client := &http.Client{}
+	resp, err := client.Do(req)
+
+	if err != nil {
+		return result, err
+	}
+
+	defer resp.Body.Close()
+
+	result, err = ioutil.ReadAll(resp.Body)
```
Comment (micahhausler) [inline pliny/pliny.go]: The following if and return are functionally unnecessary: you should be able to just 
```go
return ioutil.ReadAll(resp.Body)
```

File: types/skuid.go [hunk 5]
```diff
@@ -13,7 +13,7 @@ import (
 	"path/filepath"
 )
 
-type PullResponse struct {
+type PullResponse struct {
```
Comment (micahhausler) [inline types/skuid.go]: `go fmt` should clean this up for you, can you run `go fmt ./...`  on the project?

File: cmd/retrieve.go [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+	Short: "Retrieve Skuid metadata from a Site into a local directory.",
+	Long: "Retrieve Skuid metadata from a Skuid Platform Site and output it into a local directory.",
+	Run: func(cmd *cobra.Command, args []string) {
+
+		if verbose {
+			fmt.Println("args")
+			fmt.Println(args)
+		}
+
+		//login to a Skuid Platform Site
```
Comment (micahhausler) [inline cmd/retrieve.go]: This comment could probably go away: `.Login` is fairly self explanatory

File: cmd/retrieve.go [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+		if verbose {
+			fmt.Println("Creating file: " + file.Name)
+		}
+
+		fileReader, err := file.Open()
+		if err != nil {
+			return err
+		}
+
+		defer fileReader.Close()
```
Comment (micahhausler) [inline cmd/retrieve.go]: How many files might be opened here? 10? 100? 1000? You don't want to overrun the open file limit for the user.


`defer` is _almost_ always the way to go, but if you have potentially a ton of files, none will be closed until the end of the function, not the next iteration of the loop. (see the [go tour](https://tour.golang.org/flowcontrol/13)) You can manually just call` fileReader.Close()`  when you're done with the specific file in the loop.

Comment (micahhausler) [inline cmd/retrieve.go]: Also: since these are _written_ files, not read, you probably want to handle the possibility of an error on the `.Close()`. 

If you want to read more about this, check out this post https://joeshaw.org/dont-defer-close-on-writable-files/

File: cmd/retrieve.go [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+	Long: "Retrieve Skuid metadata from a Skuid Platform Site and output it into a local directory.",
+	Run: func(cmd *cobra.Command, args []string) {
+
+		if verbose {
+			fmt.Println("args")
+			fmt.Println(args)
+		}
+
+		//login to a Skuid Platform Site
+		api, err := pliny.Login(
```
Comment (zachelrath) [inline cmd/retrieve.go]: Oops - didn't realize this was public. I switched to using "platform" as the package name.

File: pliny/pliny.go [hunk 9]
```diff
@@ -0,0 +1,169 @@
+package pliny
```
Comment (zachelrath) [inline pliny/pliny.go]: done

File: cmd/root.go [hunk 10]
```diff
@@ -8,18 +8,21 @@ import (
 	"github.com/spf13/viper"
 )
 
-var sfdcUsername string
-var sfdcPassword string
 var appClientID string
 var appClientSecret string
 var apiVersion string
-var sfdcHost string
+var host string
```
Comment (zachelrath) [inline cmd/root.go]: done

File: pliny/pliny.go (hunk 2 above)
Comment (zachelrath) [inline pliny/pliny.go]: done

File: pliny/pliny.go [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+
+	return api, err
+}
+
+// Used to obtain an OAuth2 access_token
+func (conn *RestConnection) Refresh() (err error) {
+	urlValues := url.Values{}
+
+	urlValues.Set("grant_type", "password")
+	// urlValues.Set("client_id", conn.ClientId)
```
Comment (zachelrath) [inline pliny/pliny.go]: done

File: pliny/pliny.go [hunk 12]
```diff
@@ -0,0 +1,169 @@
+package pliny
+
+import (
+	"io/ioutil"
+	"net/http"
+
+	"fmt"
```
Comment (zachelrath) [inline pliny/pliny.go]: done

File: pliny/pliny.go [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+	err = conn.Refresh()
+
+	if err == nil {
+		api = &RestApi{
+			Connection: &conn,
+		}
+	}
+
+	if verboseLogs {
+		fmt.Println("Login successful! Access Token: " + conn.AccessToken)
```
Comment (zachelrath) [inline pliny/pliny.go]: done

File: pliny/pliny.go (hunk 3 above)
Comment (zachelrath) [inline pliny/pliny.go]: done

File: pliny/pliny.go (hunk 4 above)
Comment (zachelrath) [inline pliny/pliny.go]: done

File: cmd/retrieve.go (hunk 6 above)
Comment (zachelrath) [inline cmd/retrieve.go]: done

File: cmd/retrieve.go (hunk 7 above)
Comment (zachelrath) [inline cmd/retrieve.go]: Depending on the size of the site, it might be 100, might be 1000

Comment (zachelrath) [inline cmd/retrieve.go]: i am now calling `fileReader.Close()` directly at the end of the loop, and doing some error handling on `.Close()` for both the file and the fileReader

File: pliny/pliny.go [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+	}
+
+	return api, err
```
Comment (zachelrath) [inline pliny/pliny.go]: done

File: force/force.go [hunk 15]
```diff
@@ -36,12 +40,10 @@ func Login(consumerKey string, consumerSecret string, instanceUrl string, userna
...
 	err = conn.Refresh()
 
-	if err != nil {
-		return api, err
-	}
-
-	api = &RestApi{
-		Connection: &conn,
+	if err == nil {
+		api = &RestApi{
```
Comment (zachelrath) [inline force/force.go]: fixed

File: cmd/root.go [hunk 16]
```diff
@@ -34,19 +37,23 @@ func Execute() {
...
+	RootCmd.PersistentFlags().BoolVarP(&verbose, "verbose", "v", false, "Display all possible logging info")
+	
+	if username == "" {
+		username = os.Getenv("SKUID_UN")
 	}
 
-	if sfdcPassword == "" {
-		sfdcPassword = os.Getenv("SKUID_PW")
+	if password == "" {
+		password = os.Getenv("SKUID_PW")
```
Comment (zachelrath) [inline cmd/root.go]: fixed, now using a Replacer and Viper's `SetEnvKeyReplacer`

File: cmd/root.go [hunk 17]
```diff
@@ -34,19 +37,23 @@ func Execute() {
...
-	RootCmd.PersistentFlags().StringVar(&appClientID, "client-id", "", "Connected App Client ID")
-	RootCmd.PersistentFlags().StringVar(&appClientSecret, "client-secret", "", "Connected App Client Secret")
-	RootCmd.PersistentFlags().StringVar(&apiVersion, "api-version", "36.0", "Salesforce API Version")
-	RootCmd.PersistentFlags().StringVar(&sfdcHost, "host", "https://login.salesforce.com", "Salesforce Login Url")
-
-	if sfdcUsername == "" {
-		sfdcUsername = os.Getenv("SKUID_UN")
+	RootCmd.PersistentFlags().StringVar(&appClientID, "client-id", "", "OAuth Client ID")
+	RootCmd.PersistentFlags().StringVar(&appClientSecret, "client-secret", "", "OAuth Client Secret")
+	RootCmd.PersistentFlags().StringVar(&apiVersion, "api-version", "", "API Version")
```
Comment (zachelrath) [inline cmd/root.go]: i don't think this is necessary - you're already using environment variables to control this, no need for multiple flags in my opinion

File: pliny/pliny.go (hunk 3 above)
Comment (zachelrath) [inline pliny/pliny.go]: actually, why is this necessary?

Comment (micahhausler) [inline pliny/pliny.go]: Because the payload you're submitting is always marshaled to JSON, so it _must_ satisfy the `json.Marshaler` interface. That way you'll get a compiler error if you pass in the wrong type instead of having to check for type errors in your function. Does that make sense?

Comment (micahhausler) [inline pliny/pliny.go]: Yea you're right. https://play.golang.org/p/6rVXFabVYa

File: force/force.go [hunk 18]
```diff
@@ -1,17 +1,13 @@
 package force
 
 import (
+	"bytes"
+	"encoding/json"
+	"github.com/jpmonette/goforce"
```
Comment (micahhausler) [inline force/force.go]: Generally stdlib imports are grouped above 3rd party imports. 
```
import(
    "fmt"
    "os"
    
    "github.com/user/proj"
)

Comment (zachelrath) [inline force/force.go]: will do

File: cmd/retrieve.go [hunk 19]
```diff
@@ -0,0 +1,245 @@
...
+
+		camelCasings := make(map[string]string)
+
+		// Put in default camel casings into the map
+		for _, typeName := range allTypes {
+			camelCasings[typeName] = typeName
+		}
+		// Add in camel-casing exceptions
+		camelCasings["authproviders"] = "authProviders"
+		camelCasings["datasources"] = "dataSources"
```
Comment (zachelrath) [inline cmd/retrieve.go]: done

File: cmd/retrieve.go [hunk 20]
```diff
@@ -0,0 +1,245 @@
...
+		fetchAllByType := make(map[string]bool)
+		// Not implemented yet by Retrieve API - waiting on a PR to be merged
+		// fetchByType := make(map[string]string)
+
+		// If we have args, build up our list of args to retrieve
+		if len(args) > 0 {
+
+			for _, path := range args {
+				// Handle global wildcard
+				if strings.Contains(path, "**/*") {
```
Comment (zachelrath) [inline cmd/retrieve.go]: i have eliminated basically all of this function

File: cmd/retrieve.go [hunk 21]
```diff
@@ -0,0 +1,245 @@
...
+
+	// If we have a non-empty target directory, ensure it exists
+	if targetLocation != "" {
+		if err := os.MkdirAll(targetLocation, 0755); err != nil {
+			return err
+		}
+	}
+
+	for _, file := range reader.File {
+		path := filepath.Join(targetLocation, file.Name)
```
Comment (zachelrath) [inline cmd/retrieve.go]: done

File: cmd/retrieve.go [hunk 22]
```diff
@@ -0,0 +1,245 @@
+package cmd
```
Comment (zachelrath) [inline cmd/retrieve.go]: yep

File: cmd/retrieve.go [hunk 23]
```diff
@@ -0,0 +1,245 @@
...
+
+		if err != nil {
+			fmt.Println(err.Error())
+			os.Exit(1)
+		}
+
+		// unzip the archive into the output directory
+		targetDirFriendly := targetDir
+		if targetDir == "" {
+			targetDirFriendly = "current working directory"
```
Comment (zachelrath) [inline cmd/retrieve.go]: done

File: platform/platform.go [hunk 24]
```diff
@@ -0,0 +1,142 @@
...
+
+	conn.AccessToken = result.AccessToken
+
+	return nil
+}
+
+// Executes an HTTP request
+func (conn *RestConnection) MakeRequest(method string, url string, payload interface{}) (result []byte, err error) {
+
+	endpoint := conn.Host + "/api/v" + conn.APIVersion + url
```
Comment (micahhausler) [inline platform/platform.go]: Can this be an `fmt.Sprintf()`?

File: cmd/root.go [hunk 25]
```diff
@@ -34,19 +39,27 @@ func Execute() {
...
-	RootCmd.PersistentFlags().StringVar(&apiVersion, "api-version", "36.0", "Salesforce API Version")
-	RootCmd.PersistentFlags().StringVar(&sfdcHost, "host", "https://login.salesforce.com", "Salesforce Login Url")
+	RootCmd.PersistentFlags().StringVar(&appClientID, "client-id", "", "OAuth Client ID")
+	RootCmd.PersistentFlags().StringVar(&appClientSecret, "client-secret", "", "OAuth Client Secret")
+	RootCmd.PersistentFlags().StringVar(&apiVersion, "api-version", "", "API Version")
+	RootCmd.PersistentFlags().StringVar(&host, "host", "",
+		"API Host base URL, e.g. my-site.skuidsite.com for Skuid Platform or my-domain.my.salesforce.com for Salesforce")
+	RootCmd.PersistentFlags().StringVarP(&module, "module", "m", "", "Module name(s), separated by a comma.")
+	RootCmd.PersistentFlags().StringVarP(&password, "password", "p", "", "Skuid Platform / Salesforce Password")
+	RootCmd.PersistentFlags().StringVarP(&targetDir, "dir", "d", "", "Input/output directory.")
```
Comment (ethanfrogers) [inline cmd/root.go]: The third parameter is the default. I noticed you had a place where you checked if it was `""`. If you set it here, you won't need to do that.

File: cmd/retrieve.go [hunk 26]
```diff
@@ -0,0 +1,180 @@
...
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"strings"
+
+	"archive/zip"
+	"io/ioutil"
```
Comment (micahhausler) [inline cmd/retrieve.go]: Small bit: sorry I wasn't specific: all stdlib imports are typically grouped together. I _think_ `go fmt ./...` will do this for you

File: platform/platform.go (hunk 24 above)
Comment (ethanfrogers) [inline platform/platform.go]: I think `conn.APIVersion` should include the `v`. That looks kind of weird.

File: platform/platform.go [hunk 27]
```diff
@@ -0,0 +1,142 @@
...
+
+// Logs a given user into a target Skuid Platform site and returns a RestApi connection
+// that can be used to make HTTP requests
+func Login(host string, username string, password string, apiVersion string, verboseLogs bool) (api *RestApi, err error) {
+
+	if apiVersion == "" {
+		apiVersion = "1"
+	}
+
+	if verboseLogs {
```
Comment (ethanfrogers) [inline platform/platform.go]: Shouldn't this be `if verbose`?

File: cmd/retrieve.go (hunk 26 above)
Comment (zachelrath) [inline cmd/retrieve.go]: `go fmt ./...` has no problem with the syntax as it is here

File: platform/platform.go (hunk 24 above)
Comment (zachelrath) [inline platform/platform.go]: it app depends on how you want to pass in the API version. i.e. do we want `skuid retrieve --api-version=1` or `skuid retrieve --api-version=v1`

File: platform/platform.go (hunk 27 above)
Comment (zachelrath) [inline platform/platform.go]: sure, i changed the variable name

File: platform/platform.go (hunk 24 above)
Comment (zachelrath) [inline platform/platform.go]: @micahhausler i changed to use `fmt.Sprintf()` but honestly in my opnion string concatenation is way, way more readable. the whole `%s` thing is hideous, IMO.

**Comment by zachelrath** [comment]
All additional changes up to this point have been addressed.
